ts/packages/anchor/src/coder/borsh/event.ts

```diff
@@ -42,22 +59,36 @@ export class BorshEventCoder implements EventCoder {
         ev.name,
       ])
     );
+
+    this.address = new PublicKey(idl.address);
   }
 
+  /**
+   *
+   * @param log base 64 encoded log data from transaction message
```

Comment:
You can split this method into two: one `decode` that keeps parsing string base64 logs, and a new `decodeCpi` that accepts `base58` log from the CPI. Then no need to guess the encoding and do try-catch